--- a/ecm_db/ecm_db_mapping.h
+++ b/ecm_db/ecm_db_mapping.h
@@ -109,1 +109,12 @@
 int ecm_db_mapping_deref(struct ecm_db_mapping_instance *mi);
+#ifdef ECM_DB_ADVANCED_STATS_ENABLE
+void ecm_db_mapping_data_stats_get(struct ecm_db_mapping_instance *mi,
+                                    uint64_t *from_data_total,
+                                    uint64_t *to_data_total,
+                                    uint64_t *from_packet_total,
+                                    uint64_t *to_packet_total,
+                                    uint64_t *from_data_total_dropped,
+                                    uint64_t *to_data_total_dropped,
+                                    uint64_t *from_packet_total_dropped,
+                                    uint64_t *to_packet_total_dropped);
+#endif
