--- a/ecm_db/ecm_db_host.h
+++ b/ecm_db/ecm_db_host.h
@@ -112,1 +112,12 @@
 int ecm_db_host_deref(struct ecm_db_host_instance *hi);
+#ifdef ECM_DB_ADVANCED_STATS_ENABLE
+void ecm_db_host_data_stats_get(struct ecm_db_host_instance *hi,
+                                uint64_t *from_data_total,
+                                uint64_t *to_data_total,
+                                uint64_t *from_packet_total,
+                                uint64_t *to_packet_total,
+                                uint64_t *from_data_total_dropped,
+                                uint64_t *to_data_total_dropped,
+                                uint64_t *from_packet_total_dropped,
+                                uint64_t *to_packet_total_dropped);
+#endif
