--- a/nss_core.h
+++ b/nss_core.h
@@ -112,16 +112,1 @@
-static inline void nss_core_dma_cache_maint(void *start, uint32_t size, int direction)
-{
-	switch (direction) {
-	case DMA_FROM_DEVICE:/* invalidate only */
-		dmac_flush_range(start, start + size);
-		break;
-	case DMA_TO_DEVICE:/* writeback only */
-		dmac_flush_range(start, start + size);
-		break;
-	case DMA_BIDIRECTIONAL:/* writeback and invalidate */
-		dmac_flush_range(start, start + size);
-		break;
-	default:
-		BUG();
-	}
-}
+extern void nss_core_dma_cache_maint(void *start, uint32_t size, int direction);
--- a/nss_core.c
+++ b/nss_core.c
@@ -28,1 +28,12 @@
 #include <net/dst.h>
+
+void nss_core_dma_cache_maint(void *start, uint32_t size, int direction)
+{
+	dma_addr_t dma_addr;
+	struct device *dev = nss_top_main.nss[0].dev;
+
+	dma_addr = dma_map_single(dev, start, size, direction);
+	if (dma_mapping_error(dev, dma_addr))
+		return;
+	dma_unmap_single(dev, dma_addr, size, direction);
+}
@@ -1558,1 +1558,1 @@
-		dmac_flush_range((void *)&desc_ring[start], (void *)&desc_ring[end] + sizeof(struct n2h_descriptor));
+		NSS_CORE_DMA_CACHE_MAINT((void *)&desc_ring[start], (end - start + 1) * sizeof(struct n2h_descriptor), DMA_FROM_DEVICE);
@@ -1564,1 +1564,1 @@
-		dmac_flush_range((void *)&desc_ring[start], (void *)&desc_ring[mask] + sizeof(struct n2h_descriptor));
+		NSS_CORE_DMA_CACHE_MAINT((void *)&desc_ring[start], (mask - start + 1) * sizeof(struct n2h_descriptor), DMA_FROM_DEVICE);
@@ -1565,1 +1565,1 @@
-		dmac_flush_range((void *)&desc_ring[0], (void *)&desc_ring[end] + sizeof(struct n2h_descriptor));
+		NSS_CORE_DMA_CACHE_MAINT((void *)&desc_ring[0], (end + 1) * sizeof(struct n2h_descriptor), DMA_FROM_DEVICE);
@@ -2010,1 +2010,1 @@
-		dmac_flush_range((void *)&desc_ring[start], (void *)&desc_ring[prev_hlos_index] + sizeof(struct h2n_descriptor));
+		NSS_CORE_DMA_CACHE_MAINT((void *)&desc_ring[start], (prev_hlos_index - start + 1) * sizeof(struct h2n_descriptor), DMA_TO_DEVICE);
@@ -2016,1 +2016,1 @@
-		dmac_flush_range((void *)&desc_ring[start], (void *)&desc_ring[mask] + sizeof(struct h2n_descriptor));
+		NSS_CORE_DMA_CACHE_MAINT((void *)&desc_ring[start], (mask - start + 1) * sizeof(struct h2n_descriptor), DMA_TO_DEVICE);
@@ -2017,1 +2017,1 @@
-		dmac_flush_range((void *)&desc_ring[0], (void *)&desc_ring[prev_hlos_index] + sizeof(struct h2n_descriptor));
+		NSS_CORE_DMA_CACHE_MAINT((void *)&desc_ring[0], (prev_hlos_index + 1) * sizeof(struct h2n_descriptor), DMA_TO_DEVICE);
--- a/nss_profiler.c
+++ b/nss_profiler.c
@@ -204,1 +204,1 @@
-	dmac_flush_range(ctrl, &ctrl->cidx);
+	NSS_CORE_DMA_CACHE_MAINT(ctrl, offsetof(struct nss_profile_sdma_ctrl, cidx) + sizeof(ctrl->cidx), DMA_FROM_DEVICE);
